[Automation] Adding new API spec for 2018-01-15 version

vrdmr opened the pull request:
Added spec for new 2018-01 APIs. These mainly contain spec for the DSC APIs fixing the linting issues. Also added examples, and did some housekeeping for cleaning up the existing specs and better naming things.

----
This checklist is used to make sure that common issues in a pull request are addressed. This will expedite the process of getting your pull request merged and avoid extra work on your part to fix issues discovered during the review process.

### PR information
- [x] The title of the PR is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).
- [x] Except for special cases involving multiple contributors, the PR is started from a fork of the main repository, not a branch.
- [ ] If applicable, the PR references the bug/issue that it fixes.
- [x] Swagger files are correctly named (e.g. the `api-version` in the path should match the `api-version` in the spec).

### Quality of Swagger
- [x] I have read the [contribution guidelines](https://github.com/Azure/azure-rest-api-specs/blob/master/.github/CONTRIBUTING.md).
- [x] My spec meets the review criteria:
  - [x] The spec conforms to the [Swagger 2.0 specification](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md).
  - [x] The spec follows the guidelines described in the [Swagger checklist](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md) document.
  - [x] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.

AutorestCI commented:
# Automation for azure-sdk-for-node

AutorestCI commented:
# Automation for azure-libraries-for-java

AutorestCI commented:
# Automation for azure-sdk-for-python

AutorestCI commented:
# Automation for azure-sdk-for-go

vrdmr commented:
Depends on #2705. The refactoring is needed to go in, to fix the linting issue entirely.

mcardosos commented on specification/automation/resource-manager/readme.md:
````diff
@@ -92,10 +89,43 @@ input-file:
 - Microsoft.Automation/preview/2017-05-15-preview/job.json
 ```
 
+### Tag: package-2018-01
````
>package-2018-01 [](start = 9, length = 15)

Since this package also has files from preview versions, it should also be marked as a preview package

mcardosos replied:
Fixed

mcardosos commented on specification/automation/resource-manager/Microsoft.Automation/preview/2017-05-15-preview/definitions.json:
```diff
@@ -28,30 +28,173 @@
+        },
+        "day": {
+          "type": "string",
+          "description": "Day of the occurrence. Must be one of monday, tuesday, wednesday,thursday, friday, saturday, sunday.",
```
>thursday [](start = 91, length = 8)

Super nit, space

mcardosos replied:
Fixed

mcardosos commented on specification/automation/resource-manager/Microsoft.Automation/preview/2017-05-15-preview/definitions.json:
```diff
@@ -28,30 +28,173 @@
+    },
+    "AdvancedSchedule": {
+      "properties": {
+        "weekDays": {
```
>weekDays [](start = 9, length = 8)

Are the possible values the same as in line 44?

mcardosos replied:
Okok

mcardosos commented on specification/automation/resource-manager/Microsoft.Automation/preview/2017-05-15-preview/definitions.json:
```diff
@@ -28,30 +28,173 @@
+      },
+      "description": "Definition of job properties."
     },
     "jobCollectionItem": {
```
>jobCollectionItem [](start = 5, length = 17)

Should this be PascalCase?

vrdmr replied:
Done

mcardosos commented on specification/automation/resource-manager/Microsoft.Automation/preview/2017-05-15-preview/definitions.json:
```diff
@@ -1051,9 +1384,156 @@
+          "format": "double",
+          "description": "Gets or sets the next run time's offset in minutes."
+        },
+        "interval": {
```
>interval [](start = 9, length = 8)

What type is this property?

vrdmr replied:
Checking with the JobService team. This will go in a separate JobService PR. (mentioned above)

vrdmr replied:
Fixed.

mcardosos commented on specification/automation/resource-manager/Microsoft.Automation/preview/2017-05-15-preview/definitions.json:
```diff
@@ -1051,9 +1384,156 @@
+    "JobCreateProperties": {
+      "properties": {
+        "runbook": {
+          "$ref": "./definitions.json#/definitions/RunbookAssociationProperty",
```
>./definitions.json [](start = 19, length = 18)

This is already the definiitions file

mcardosos replied:
Fixed

mcardosos commented on specification/automation/resource-manager/Microsoft.Automation/preview/2017-05-15-preview/definitions.json:
```diff
@@ -1051,9 +1384,156 @@
+      "type": "string",
+      "required": true,
+      "in": "path",
+      "x-ms-parameter-location": "client"
```
>client [](start = 34, length = 6)

This goes better in the method instead of the client

mcardosos replied:
Fixed

mcardosos commented on specification/automation/resource-manager/Microsoft.Automation/stable/2018-01-15/definitions.json:
```diff
@@ -0,0 +1,147 @@
+      "type": "string",
+      "required": true,
+      "in": "path",
+      "x-ms-parameter-location": "client"
```
>client [](start = 34, length = 6)

This goes better in the method

vrdmr replied:
Done.

mcardosos replied:
Err, it should be `"x-ms-parameter-location": "method"`

---
In reply to: [175917693](https://github.com/Azure/azure-rest-api-specs/pull/2690#discussion_r175917693) [](ancestors = 175917693)

vrdmr replied:
Done

mcardosos replied:
Fixed

mcardosos commented on specification/automation/resource-manager/Microsoft.Automation/stable/2018-01-15/dscCompilationJob.json:
```diff
@@ -0,0 +1,608 @@
+        "tags": [
+          "DscCompilationJob"
+        ],
+        "operationId": "DscCompilationJob_Create",
```
>DscCompilationJob_Create [](start = 24, length = 24)

Is this a long running operation?

vrdmr replied:
No this is not. It just submits a job and returns the JobId, which is then used to track the progress of the Job.

mcardosos replied:
Should clients then poll for operation completion?

vrdmr replied:
Yes.

mcardosos replied:
Looks like a long running operation to me

vrdmr replied:
After [reading this](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/creating-swagger.md#enable-asynchronous-operations-with-x-ms-long-running-operation) up, it makes sense to mark this as ` x-ms-long-running-operation`. I will do that.

vrdmr replied:
Done.

mcardosos replied:
Fixed

mcardosos commented on specification/automation/resource-manager/Microsoft.Automation/stable/2018-01-15/dscCompilationJob.json:
```diff
@@ -0,0 +1,608 @@
+            "$ref": "./definitions.json#/parameters/AutomationAccountNameParameter"
+          },
+          {
+            "name": "$filter",
```
>$filter [](start = 21, length = 7)

x-ms-odata extension would be valuabe for this operation

vrdmr replied:
Thanks. I'll talk with the team and take this up as a separate PR as it is not a blocker right now.

mcardosos replied:
As this is a comment for a swagger in the stable folder, I strongly recommend to get whatever is inside this folder right in one PR. This way you will avoid breaking changes on generated SDKs.

vrdmr replied:
Done.

mcardosos replied:
Fixed

mcardosos commented on specification/automation/resource-manager/Microsoft.Automation/stable/2018-01-15/dscCompilationJob.json:
```diff
@@ -0,0 +1,608 @@
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Automation/automationAccounts/{automationAccountName}/compilationjobs/{jobId}/streams/": {
```
>/ [](start = 177, length = 1)

Is this last slash really needed?

vrdmr replied:
Done.

mcardosos replied:
It is still there :(

vrdmr replied:
oh. It could be because of the merge issues. Fixing it now.

mcardosos replied:
Fixed

mcardosos commented on specification/automation/resource-manager/Microsoft.Automation/stable/2018-01-15/dscNode.json:
```diff
@@ -0,0 +1,957 @@
+          "type": "string",
+          "description": "Gets or sets the name of the resource."
+        },
+        "location": {
```
>location [](start = 9, length = 8)

So, when regenerating keys, could the location also be updated?

vrdmr replied:
Checking with the team.

vrdmr replied:
Fixed.

mcardosos commented on specification/automation/resource-manager/Microsoft.Automation/stable/2018-01-15/dscNode.json:
```diff
@@ -0,0 +1,957 @@
+          "type": "string",
+          "description": "Gets or sets the account id of the node."
+        },
+        "nodeConfiguration": {
```
>nodeConfiguration [](start = 9, length = 17)

WDYT about flattening this property?

vrdmr replied:
Could you describe a little more about this? You mean the x-ms-client-flatten property?

mcardosos replied:
yes

vrdmr replied:
Done.

mcardosos replied:
Fixed

mcardosos commented on specification/automation/resource-manager/Microsoft.Automation/stable/2018-01-15/dscNode.json:
```diff
@@ -0,0 +1,957 @@
+        "tags": [
+          "DscNode"
+        ],
+        "operationId": "DscNode_ListByAutomationAccount",
```
>DscNode_ListByAutomationAccount [](start = 24, length = 31)

This operation could also use x-ms-odata

vrdmr replied:
Same as above: Thanks. I'll talk with the team and take this up as a separate PR as it is not a blocker right now.

vrdmr replied:
Done

mcardosos replied:
Fixed

mcardosos commented on specification/automation/resource-manager/Microsoft.Automation/stable/2018-01-15/dscNode.json:
```diff
@@ -0,0 +1,957 @@
+        "tags": [
+          "NodeReports"
+        ],
+        "operationId": "NodeReports_ListByNode",
```
>NodeReports_ListByNode [](start = 24, length = 22)

This one too, x-ms-odata

vrdmr replied:
Same as above: Thanks. I'll talk with the team and take this up as a separate PR as it is not a blocker right now.

vrdmr replied:
Done

mcardosos replied:
Fixed

mcardosos commented on specification/automation/resource-manager/Microsoft.Automation/stable/2018-01-15/dscNode.json:
```diff
@@ -0,0 +1,957 @@
+          "type": "string",
+          "description": "Gets or sets the status of the node report."
+        },
+        "refreshMode": {
```
>refreshMode [](start = 9, length = 11)

Could this be an enum?

vrdmr replied:
We aren't planning to enum this currently as the things being returned here are not finalized yet and are subject to change.

mcardosos commented on specification/automation/resource-manager/Microsoft.Automation/stable/2018-01-15/dscNode.json:
```diff
@@ -0,0 +1,957 @@
+      },
+      "description": "Definition of the DSC Meta Configuration."
+    },
+    "DscNodeReport": {
```
>DscNodeReport [](start = 5, length = 13)

Some properties in this definition, I believe could be represented with enums

vrdmr replied:
Yup. Agreed. Let me check the implementation to verify.

vrdmr replied:
Same as above: We aren't planning to enum this currently as the things being returned here are not finalized yet and are subject to change.

mcardosos commented on specification/automation/resource-manager/Microsoft.Automation/stable/2018-01-15/dscNode.json:
```diff
@@ -0,0 +1,957 @@
+          "200": {
+            "description": "OK",
+            "schema": {
+              "type": "file"
```
>file [](start = 23, length = 4)

This is not what I see in the report. It looks like this could be defined as a model

vrdmr replied:
I think the example is wrong in this case. It does not return a well-formed JSON

It sends something like this: 
URL: `https://management.azure.com/subscriptions/b5e4748c-f69a-467c-8749-e2f9c8cd3db0/resourceGroups/ConfigDataValidation/providers/Microsoft.Automation/automationAccounts/ConfigDataValidation/nodes/119a0b96-23fd-11e8-b29c-000d3a18dec4/reports/76f09587-2c84-11e8-a9ca-000d3a18dec4/content?api-version=2018-01-15`

The Response - which is raw reports from our DSC Nodes.
```
{"JobId":"76f09587-2c84-11e8-a9ca-000d3a18dec4","OperationType":"Consistency","NodeName":"ANAGG-2008R2","IpAddress":"10.13.49.8;127.0.0.1;fe80::2cc0:8062:a210:e1c6%11;::2000:0:0:0;::1;::2000:0:0:0","LCMVersion":"2.0","ReportFormatVersion":"2.0","StartTime":"2018-03-20T21:20:09.0080000+00:00","Errors":[],"StatusData":["{\"IPV4Addresses\":[\"10.13.49.8\",\"127.0.0.1\"],\"MACAddresses\":[\"00-0D-3A-18-DE-C4\",\"00-00-00-00-00-00-00-E0\"],\"Type\":\"Consistency\",\"HostName\":\"ANAGG-2008R2\",\"Locale\":\"en-US\",\"StartDate\":\"2018-03-20T21:20:09.0080000+00:00\",\"JobID\":\"{76F09587-2C84-11E8-A9CA-000D3A18DEC4}\",\"LCMVersion\":\"2.0\",\"IPV6Addresses\":[\"fe80::2cc0:8062:a210:e1c6%11\",\"::2000:0:0:0\",\"::1\",\"::2000:0:0:0\"]}"],"AdditionalData":[{"Key":"OSVersion","Value":"{\"VersionString\":\"Microsoft Windows NT 6.1.7601 Service Pack 1\",\"ServicePack\":\"Service Pack 1\",\"Platform\":\"Win32NT\"}"},{"Key":"PSVersion","Value":"{\"CLRVersion\":\"4.0.30319.42000\",\"PSVersion\":\"5.1.14409.1012\",\"BuildVersion\":\"10.0.14409.1012\"}"}]}

{"JobId":"76f09587-2c84-11e8-a9ca-000d3a18dec4","OperationType":"Consistency","RefreshMode":"Pull","Status":"Success","ReportFormatVersion":"2.0","StartTime":"2018-03-20T21:20:09.0080000+00:00","EndTime":"2018-03-20T21:20:37.0080000+00:00","RebootRequested":"False","Errors":[],"StatusData":["{\"StartDate\":\"2018-03-20T21:20:09.0080000+00:00\",\"IPV6Addresses\":[\"fe80::2cc0:8062:a210:e1c6%11\",\"::2000:0:0:0\",\"::1\",\"::2000:0:0:0\"],\"MACAddresses\":[\"00-0D-3A-18-DE-C4\",\"00-00-00-00-00-00-00-E0\"],\"DurationInSeconds\":\"28\",\"Status\":\"Success\",\"IPV4Addresses\":[\"10.13.49.8\",\"127.0.0.1\"],\"LCMVersion\":\"2.0\",\"Type\":\"Consistency\",\"HostName\":\"ANAGG-2008R2\",\"RebootRequested\":\"False\",\"JobID\":\"{76F09587-2C84-11E8-A9CA-000D3A18DEC4}\",\"MetaConfiguration\":{\"AgentId\":\"119A0B96-23FD-11E8-B29C-000D3A18DEC4\",\"SignatureValidations\":[],\"ConfigurationDownloadManagers\":[{\"RegistrationKey\":\"\",\"ServerURL\":\"https://eus2-agentservice-prod-1.azure-automation.net/accounts/bd8fac9e-01c5-4523-98e0-f273f474fbf6\",\"ResourceId\":\"[ConfigurationRepositoryWeb]AzureAutomationDSC\",\"ConfigurationNames\":[],\"SourceInfo\":\"C:\\\\Packages\\\\Plugins\\\\Microsoft.Powershell.DSC\\\\2.75.0.0\\\\DSCWork\\\\RegistrationMetaConfigV2.0\\\\RegistrationMetaConfigV2.ps1::69::9::ConfigurationRepositoryWeb\"}],\"ActionAfterReboot\":\"ContinueConfiguration\",\"LCMCompatibleVersions\":[\"1.0\",\"2.0\"],\"LCMState\":\"Idle\",\"ResourceModuleManagers\":[{\"RegistrationKey\":\"\",\"ServerURL\":\"https://eus2-agentservice-prod-1.azure-automation.net/accounts/bd8fac9e-01c5-4523-98e0-f273f474fbf6\",\"ResourceId\":\"[ResourceRepositoryWeb]AzureAutomationDSC\",\"SourceInfo\":\"C:\\\\Packages\\\\Plugins\\\\Microsoft.Powershell.DSC\\\\2.75.0.0\\\\DSCWork\\\\RegistrationMetaConfigV2.0\\\\RegistrationMetaConfigV2.ps1::76::9::ResourceRepositoryWeb\"}],\"ReportManagers\":[{\"RegistrationKey\":\"\",\"ServerURL\":\"https://eus2-agentservice-prod-1.azure-automation.net/accounts/bd8fac9e-01c5-4523-98e0-f273f474fbf6\",\"ResourceId\":\"[ReportServerWeb]AzureAutomationDSC\",\"SourceInfo\":\"C:\\\\Packages\\\\Plugins\\\\Microsoft.Powershell.DSC\\\\2.75.0.0\\\\DSCWork\\\\RegistrationMetaConfigV2.0\\\\RegistrationMetaConfigV2.ps1::82::9::ReportServerWeb\"}],\"StatusRetentionTimeInDays\":\"10\",\"LCMVersion\":\"2.0\",\"MaximumDownloadSizeMB\":\"500\",\"ConfigurationMode\":\"ApplyAndMonitor\",\"RefreshFrequencyMins\":\"30\",\"RebootNodeIfNeeded\":\"False\",\"SignatureValidationPolicy\":\"NONE\",\"RefreshMode\":\"Pull\",\"DebugMode\":[\"NONE\"],\"CertificateID\":\"24C07ACB944D2781BDB53E2CB636CEE890BDF39E\",\"AllowModuleOverwrite\":\"False\",\"ConfigurationModeFrequencyMins\":\"15\",\"LCMStateDetail\":\"\"},\"Locale\":\"en-US\",\"Mode\":\"Pull\"}"],"AdditionalData":[{"Key":"OSVersion","Value":"{\"VersionString\":\"Microsoft Windows NT 6.1.7601 Service Pack 1\",\"ServicePack\":\"Service Pack 1\",\"Platform\":\"Win32NT\"}"},{"Key":"PSVersion","Value":"{\"CLRVersion\":\"4.0.30319.42000\",\"PSVersion\":\"5.1.14409.1012\",\"BuildVersion\":\"10.0.14409.1012\"}"}]}

```

mcardosos replied:
Still looks like something that can be modeled

vrdmr replied:
We don't want to model this currently. As this is subject to change based on what the agent sends and we don't want this to be a breaking change. Please let me know if this is a blocker for this PR.

mcardosos replied:
What about returning a model with `additionalProperties` ?

mcardosos commented on specification/automation/resource-manager/Microsoft.Automation/stable/2018-01-15/dscNodeConfiguration.json:
```diff
@@ -0,0 +1,416 @@
+      },
+      "description": "The response model for the list job operation."
+    },
+    "DscNodeConfigurationAssociationProperty": {
```
>DscNodeConfigurationAssociationProperty [](start = 5, length = 39)

This definition is duplicated (it lives both in `dscNode` and this file). Would it be better to define it only one in the `definitions` file?

mcardosos replied:
Actually, this model is not used at all in this file. It is best to remove it.

---
In reply to: [175907802](https://github.com/Azure/azure-rest-api-specs/pull/2690#discussion_r175907802) [](ancestors = 175907802)

vrdmr replied:
I did not do that as I was facing other issues related to DSCNode issue.  I'll take up the refactoring task after the change in the old spec are in.

```
ERROR (BodyTopLevelProperties/R3006/RPCViolation): Top level properties should be one of name, type, id, location, properties, tags, plan, sku, etag, managedBy, identity, zones. Model definition 'DscNode' has extra properties ['lastSeen,registrationTime,ip,accountId,nodeConfiguration,status,nodeId,extensionHandler'].
    - file:///Users/varadmeru/work/microsoft/github/azure-rest-api-specs/specification/automation/resource-manager/Microsoft.Automation/stable/2015-10-31/definitions.json:1463:6 ($.definitions.DscNode.properties)

ERROR (BodyTopLevelProperties/R3006/RPCViolation): Top level properties should be one of name, type, id, location, properties, tags, plan, sku, etag, managedBy, identity, zones. Model definition 'DscNodeConfiguration' has extra properties ['lastModifiedTime,creationTime,configuration'].
    - file:///Users/varadmeru/work/microsoft/github/azure-rest-api-specs/specification/automation/resource-manager/Microsoft.Automation/stable/2015-10-31/definitions.json:1539:6 ($.definitions.DscNodeConfiguration.properties)
```

vrdmr replied:
Discard the above comment. I had not understand your comment. Done. Removed.

mcardosos replied:
Fixed

mcardosos commented on specification/automation/resource-manager/Microsoft.Automation/stable/2018-01-15/dscNodeConfiguration.json:
```diff
@@ -0,0 +1,416 @@
+        "tags": [
+          "DscNodeConfiguration"
+        ],
+        "operationId": "DscNodeConfiguration_CreateOrUpdate",
```
>DscNodeConfiguration_CreateOrUpdate [](start = 24, length = 35)

Is this a long running operation?

vrdmr replied:
No. DscNodeconfigurations are usually very small in size. and most of the PUT calls use bloblink rather than upload from local files.

mcardosos replied:
Does the client have to poll?

vrdmr replied:
Yes

vrdmr replied:
Done.

mcardosos replied:
Fixed

mcardosos commented on specification/automation/resource-manager/Microsoft.Automation/stable/2018-01-15/dscNodeConfiguration.json:
```diff
@@ -0,0 +1,416 @@
+        "tags": [
+          "DscNodeConfiguration"
+        ],
+        "operationId": "DscNodeConfiguration_ListByAutomationAccount",
```
>DscNodeConfiguration_ListByAutomationAccount [](start = 24, length = 44)

x-ms-odata can be used here

vrdmr replied:
Same as above.

vrdmr replied:
Done

mcardosos commented:
Closing and reopening to trigger CI

mcardosos commented on specification/automation/resource-manager/Microsoft.Automation/stable/2018-01-15/definitions.json:
```diff
@@ -0,0 +1,147 @@
+      },
+      "description": "Error response of an operation failure"
+    },
+    "Resource": {
```
Could this pattern be followed?
```json
"Resource": {
      "description": "The core properties of ARM resources",
      "properties": {
        "id": {
          "readOnly": true,
          "type": "string",
          "description": "Fully qualified resource Id for the resource"
        },
        "name": {
          "readOnly": true,
          "type": "string",
          "description": "The name of the resource"
        },
        "type": {
          "readOnly": true,
          "type": "string",
          "description": "The type of the resource."
        }
      },
      "x-ms-azure-resource": true
    },
    "TrackedResource": {
      "description": "The resource model definition for a ARM tracked top level resource",
      "properties": {
        "tags": {
          "type": "object",
          "additionalProperties": {
            "type": "string"
          },
          "x-ms-mutability": [
            "read",
            "create",
            "update"
          ],
          "description": "Resource tags."
        },
        "location": {
          "type": "string",
          "x-ms-mutability": [
            "read",
            "create"
          ],
          "description": "The Azure Region where the resource lives"
        }
      },
      "allOf": [
        {
          "$ref": "#/definitions/Resource"
        }
      ]
    },
    "ProxyResource": {
      "description": "The resource model definition for a ARM proxy resource. It will have everything other than required location and tags",
      "allOf": [
        {
          "$ref": "#/definitions/Resource"
        }
      ]
    },

```

vrdmr replied:
Done.

mcardosos replied:
Fixed

mcardosos commented on specification/automation/resource-manager/Microsoft.Automation/stable/2018-01-15/dscCompilationJob.json:
```diff
@@ -0,0 +1,608 @@
+        }
+      },
+      "x-ms-client-flatten": true,
+      "description": "Definition of the job stream."
```
I dont think the extension is needed here. It is already used in the definiitions that reference this one.

vrdmr replied:
Fixed.

mcardosos replied:
Fixed

mcardosos commented on specification/automation/resource-manager/Microsoft.Automation/preview/2017-05-15-preview/definitions.json:
```diff
@@ -28,55 +28,297 @@
+        }
+      },
+      "x-ms-client-flatten": true,
+      "description": "Definition of the job stream."
```
This extension is not needed here

vrdmr replied:
Fixed.

mcardosos replied:
Fixed

mcardosos reviewed (changes requested):
Added some comments.
Also there are some [linter errors](https://travis-ci.org/Azure/azure-rest-api-specs/jobs/356052336#L778) and [model validation errors](https://travis-ci.org/Azure/azure-rest-api-specs/jobs/356052337#L760) to take a look at.

vrdmr commented on specification/automation/resource-manager/Microsoft.Automation/stable/2018-01-15/dscCompilationJob.json:
```diff
@@ -0,0 +1,608 @@
+          "readOnly": true,
+          "description": "Gets the creation time of the job."
+        },
+        "provisioningState": {
```
I just copied the spec from [previous stable spec](https://github.com/vrdmr/azure-rest-api-specs/blob/vameru-automation-new-api-2018-01-15/specification/automation/resource-manager/Microsoft.Automation/stable/2015-10-31/definitions.json#L1132-L1135). The provisioningState does not have provisioningState. How do you suggest to fix this? I can remove the inner provisioningState.

mcardosos replied:
The swagger should reflect faithfully the API. It just looks weird to me to have a provisiningState property inside another provisiningState property. But if that is how the service is behaving, leave it like that.

vrdmr committed:
Adding new API spec for 2018-01-15 version

vrdmr committed:
Merge branch 'master' into vameru-automation-new-api-2018-01-15

vrdmr commented:
@mcardosos Could you please take a look at the changes. Thanks.

vrdmr committed:
Fixing the Examples API version

vrdmr committed:
Merge branch 'master' into vameru-automation-new-api-2018-01-15

vrdmr committed:
Updating the sourControlSyncJobs to fix the model validation issues

vrdmr committed:
Fix the model issue with getContent

vrdmr committed:
Updated SoftwareUpdateConfiguration examples

vrdmr commented:
@mcardosos @ravbhatnagar Could you please take a look at this PR and approve it. I've fixed all the issues related to model/examples and also addressed the review comments.

mcardosos commented:
> 

Have you already requested this suppressions? Could you please add me to the email thread if so?

---
Refers to: specification/automation/resource-manager/readme.md:137 in 8b68d37. [](commit_id = 8b68d3738a2b38e865aaa3c5daa34cedcb4ace5b, deletion_comment = False)

mcardosos commented on specification/automation/resource-manager/Microsoft.Automation/preview/2017-05-15-preview/definitions.json:
```diff
@@ -1051,9 +1386,156 @@
+      "pattern": "^[-\\w\\._]+$",
+      "minLength": 1,
+      "maxLength": 90,
+      "x-ms-parameter-location": "client",
```
>client [](start = 34, length = 6)

It is preferred for resourcegroup paramter to have ` "x-ms-parameter-location": "method"`?

vrdmr replied:
@shahabhijeet and the team had a discussion if we would like to place RG and AutomationAccountName and we decided to remove the location to the client. But yes, this should be in a separate PR. I'll make this change now.

vrdmr commented:
@mcardosos - The suppression PR #2749. I've also sent you the mail.

vrdmr committed:
Change the location scope of ResourceGroupNameParam

mcardosos commented:
MIssing stuff to fix:
- Interval type.
- Extra slash.
- Additional properties model.
- Odata extension, dscNodeReport, refresh enum. Strongly recommended to include in this PR since those are changes on the stable folder.

There might still be extra stuff to fix (we will know once CI finishes)

vrdmr committed:
Fix based on recent comments

vrdmr commented:
@mcardosos Out of the above - 

> MIssing stuff to fix:
> 
> Interval type.

Will update this asap.

> Extra slash.

Removed

> Additional properties model.

As you said above, this is not a blocker currently. Also, when I remove the `provisioningState` from `JobProvisioningStateProperty`, the model issues start showing up. Also, with the current change, `oav validate-example` does not complain currently.

> Odata extension, dscNodeReport, refresh enum. Strongly recommended to include in this PR since those are changes on the stable folder.

OData extension done. After talking to the team, the dscNodeReport enums is not in plan currently.

> There might still be extra stuff to fix (we will know once CI finishes)

Will keep an eye.

vrdmr committed:
Adding type for interval

mcardosos commented on specification/automation/resource-manager/Microsoft.Automation/preview/2017-05-15-preview/definitions.json:
```diff
@@ -1051,9 +1386,158 @@
+        },
+        "interval": {
+          "type": "integer",
+          "format": "byte",
```
>byte [](start = 21, length = 4)

Does not make sense for this to be format byte, PTAL here
[https://swagger.io/docs/specification/data-models/data-types/](https://swagger.io/docs/specification/data-models/data-types/)

vrdmr replied:
Fixed. Changed it to String/Byte format (ref: http://docs.swagger.io/spec.html)

mcardosos replied:
Fixed

vrdmr committed:
Corrent the type for interval type

mcardosos commented:
Cool! So missing stuff to fix are [linter errors](https://travis-ci.org/Azure/azure-rest-api-specs/jobs/359492021#L809) and [model validation errors](https://travis-ci.org/Azure/azure-rest-api-specs/jobs/359492022#L780).

vrdmr commented:
For the model issues, I already made the fix (https://github.com/Azure/azure-rest-api-specs/pull/2690/commits/8b68d3738a2b38e865aaa3c5daa34cedcb4ace5b). Don't know why its now showing up. I'll make the commit again.

As for the linter issues. It is finding linting issues in 2015-10-31 folder, which I am fixing in 2018-01-15 resources. I am not changing previous stable version to fix it.

vrdmr committed:
Fixing model issues in SUC resources

vrdmr committed:
Fixed the interval type model issue

vrdmr commented:
@mcardosos Fixed the model issues.

mcardosos commented on specification/automation/resource-manager/readme.md:
````diff
@@ -196,6 +231,15 @@ Please also specify `--go-sdk-folder=<path to the root directory of your azure-s
+Please also specify `--go-sdk-folder=<path to the root directory of your azure-sdk-for-go clone>`.
+
+``` yaml $(tag) == 'package-2018-01-preview' && $(go)
+output-folder: $(go-sdk-folder)/services/automation/mgmt/2018-01-15-preview
````
>/ [](start = 40, length = 1)

Add a preview folder here :D

vrdmr replied:
I don't see other services adding a `preview` folder or a preview folder showing up in the Go SDK repo.

Example - [Data Migration](https://github.com/Azure/azure-rest-api-specs/tree/master/specification/compute/resource-manager#tag-package-container-service-2015-11-preview-and-go) and [Compute](https://github.com/Azure/azure-rest-api-specs/tree/master/specification/compute/resource-manager#tag-package-container-service-2015-11-preview-and-go). Do I need to add this then?

mcardosos replied:
There is not problem if the golang section is left out. We will figure it out later

vrdmr replied:
What do you suggest then, for getting the PR through? I can send a separate PR for the same.

mcardosos commented on specification/automation/resource-manager/Microsoft.Automation/stable/2015-10-31/definitions.json:
```diff
@@ -920,7 +928,11 @@
-      "x-ms-azure-resource": true,
+      "allOf": [
+        {
+          "$ref": "#/definitions/ProxyResource"
```
On the other hand this changes might be breaking changes for some SDKs. WDYT about adding these improvements into a new file that is only used by the API version introduced in this PR?

---
In reply to: [177918048](https://github.com/Azure/azure-rest-api-specs/pull/2690#discussion_r177918048) [](ancestors = 177918048,177918018)

vrdmr replied:
Yes. But but I didn't want to change the meaning behind it. This can go to the discussion why have `ProxyResource` definition at all if all the resources can be either `TrackedResource` or just `Resource`. Having `ProxyResource` makes the spec clearer and explicit.

mcardosos replied:
What about the no longer needed properties?

vrdmr commented:
@ravbhatnagar Could you please take a look at this PR. Thanks.

vrdmr committed:
Change readme.md to remove new go build

vrdmr committed:
Fix Proxy Resource Properties

vrdmr committed:
Updating DSC Compilation Jobs examples

mcardosos reviewed (approved):
Cool! :)
Linter errors are not related to this PR, but... awareness

ravbhatnagar commented on specification/automation/resource-manager/Microsoft.Automation/stable/2018-01-15/dscNode.json:
```diff
@@ -0,0 +1,949 @@
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Automation/automationAccounts/{automationAccountName}/agentRegistrationInformation/regenerateKey": {
```
URL is not formed correctly. This will have RBAC issues. Is this new API or existing? Is agentInformation a resource type defined in your manifest?

weijiel replied:
This is existing. It's a proxy resources therefore not defined in the manifest. It's used by our portal code but not through ARM template.

ravbhatnagar commented on specification/automation/resource-manager/Microsoft.Automation/stable/2018-01-15/dscNode.json:
```diff
@@ -0,0 +1,949 @@
+          "type": "string",
+          "description": "Gets or sets the dsc server endpoint."
+        },
+        "keys": {
```
keys should not be returned as part of the GET. Only via a POST.

weijiel replied:
This is also legacy API

ravbhatnagar commented on specification/automation/resource-manager/Microsoft.Automation/stable/2018-01-15/dscNode.json:
```diff
@@ -0,0 +1,949 @@
+      },
+      "description": "Definition of the dsc node report error type."
+    },
+    "AgentRegistration": {
```
If this is a resource, it is missing the name, type properties. Is the id property the ARM resource id or something else.

weijiel replied:
This is a proxy resource. Also part of existing API.

ravbhatnagar commented:
@vrdmr - Some comments. Not sure if these are existing APis or new. Please ping me to close on this.

ravbhatnagar commented:
OK since these are all existing APis, signing off despite these issues.
